File: intel/wolfpack/exchanges/dydx.py
"""dYdX v4 exchange adapter using the Indexer REST API."""

import asyncio
import logging
import time

import httpx

from wolfpack.exchanges.base import (
    ExchangeAdapter,
    ExchangeId,
    Candle,
    FundingRate,
    MarketInfo,
    Orderbook,
    OrderbookLevel,
)

logger = logging.getLogger(__name__)

BASE_URL = "https://indexer.dydx.trade/v4"

INTERVAL_MAP = {
    "1m": "1MIN",
    "5m": "5MINS",
    "15m": "15MINS",
    "30m": "30MINS",
    "1h": "1HOUR",
    "4h": "4HOURS",
    "1d": "1DAY",
}

# Simple TTL cache (matches hyperliquid.py pattern)
_cache: dict[str, tuple[float, object]] = {}
_CACHE_TTLS = {
    "markets": 300,
    "candles": 10,
    "funding": 60,
    "orderbook": 2,
}


def _cache_get(key: str) -> object | None:
    if key in _cache:
        ts, data = _cache[key]
        ttl = _CACHE_TTLS.get(key.split(":")[0], 60)
        if time.time() - ts < ttl:
            return data
    return None


def _cache_set(key: str, data: object) -> None:
    _cache[key] = (time.time(), data)


class DydxExchange(ExchangeAdapter):
    def __init__(self) -> None:
        self._client: httpx.AsyncClient | None = None

    async def _get_client(self) -> httpx.AsyncClient:
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(timeout=10)
        return self._client

    async def _get(self, path: str, params: dict | None = None) -> dict | list:
        client = await self._get_client()
        resp = await client.get(f"{BASE_URL}{path}", params=params)
        resp.raise_for_status()
        return resp.json()

    @property
    def id(self) -> ExchangeId:
        return "dydx"

    @property
    def name(self) -> str:
        return "dYdX"

    async def get_markets(self) -> list[MarketInfo]:
        cache_key = "markets"
        cached = _cache_get(cache_key)
        if cached is not None:
            return cached  # type: ignore

        data = await self._get("/perpetualMarkets")
        markets_raw = data.get("markets", {}) if isinstance(data, dict) else {}
        result = [
            MarketInfo(
                symbol=str(info.get("ticker", k)),
                base_asset=str(info.get("ticker", k)).replace("-USD", ""),
                last_price=float(info.get("oraclePrice", 0)),
                volume_24h=float(info.get("volume24H", 0)),
                open_interest=float(info.get("openInterest", 0)),
                funding_rate=0,
                max_leverage=20,
            )
            for k, info in markets_raw.items()
        ]
        _cache_set(cache_key, result)
        return result

    async def get_candles(
        self, symbol: str, interval: str = "1h", limit: int = 100, start_time: int | None = None
    ) -> list[Candle]:
        cache_key = f"candles:{symbol}:{interval}:{limit}:{start_time}"
        cached = _cache_get(cache_key)
        if cached is not None:
            return cached  # type: ignore

        resolution = INTERVAL_MAP.get(interval, "1HOUR")
        params: dict[str, str | int] = {"resolution": resolution, "limit": limit}
        if start_time is not None:
            from datetime import datetime, timezone
            params["fromISO"] = datetime.fromtimestamp(start_time / 1000, tz=timezone.utc).isoformat()
        data = await self._get(
            f"/candles/perpetualMarkets/{symbol}",
            params=params,
        )
        result = [
            Candle(
                timestamp=_parse_ts(c.get("startedAt", "")),
                open=float(c["open"]),
                high=float(c["high"]),
                low=float(c["low"]),
                close=float(c["close"]),
                volume=float(c.get("baseTokenVolume", 0)),
            )
            for c in (data.get("candles", []) if isinstance(data, dict) else [])
        ]
        _cache_set(cache_key, result)
        return result

    async def get_funding_rates(self) -> list[FundingRate]:
        cache_key = "funding"
        cached = _cache_get(cache_key)
        if cached is not None:
            return cached  # type: ignore

        markets = await self.get_markets()
        client = await self._get_client()

        # Fetch funding rates concurrently instead of sequentially
        async def _fetch_one(symbol: str) -> FundingRate:
            try:
                resp = await client.get(
                    f"{BASE_URL}/historicalFunding/{symbol}",
                    params={"limit": 1},
                    timeout=5,
                )
                data = resp.json()
                latest = (data.get("historicalFunding") or [None])[0]
                return FundingRate(
                    symbol=symbol,
                    rate=float(latest["rate"]) if latest else 0,
                    next_funding_time=_now_ms() + 3_600_000,
                )
            except Exception:
                return FundingRate(
                    symbol=symbol, rate=0, next_funding_time=_now_ms() + 3_600_000
                )

        tasks = [_fetch_one(m.symbol) for m in markets[:10]]
        result = await asyncio.gather(*tasks)
        rates = list(result)
        _cache_set(cache_key, rates)
        return rates

    async def get_orderbook(self, symbol: str, depth: int = 20) -> Orderbook:
        cache_key = f"orderbook:{symbol}"
        cached = _cache_get(cache_key)
        if cached is not None:
            return cached  # type: ignore

        data = await self._get(f"/orderbooks/perpetualMarket/{symbol}")
        result = Orderbook(
            symbol=symbol,
            bids=[
                OrderbookLevel(price=float(b["price"]), size=float(b["size"]))
                for b in (data.get("bids") or [] if isinstance(data, dict) else [])[:depth]
            ],
            asks=[
                OrderbookLevel(price=float(a["price"]), size=float(a["size"]))
                for a in (data.get("asks") or [] if isinstance(data, dict) else [])[:depth]
            ],
            timestamp=_now_ms(),
        )
        _cache_set(cache_key, result)
        return result

    async def close(self) -> None:
        if self._client and not self._client.is_closed:
            await self._client.aclose()


def _now_ms() -> int:
    return int(time.time() * 1000)


def _parse_ts(iso_str: str) -> int:
    from datetime import datetime, timezone
    try:
        return int(datetime.fromisoformat(iso_str.replace("Z", "+00:00")).timestamp() * 1000)
    except Exception:
        return _now_ms()

File: app/src/lib/exchange/dydx.ts
import type {
  ExchangeAdapter,
  ExchangeCredentials,
  MarketInfo,
  Candle,
  FundingRate,
  OrderbookSnapshot,
  AccountBalance,
  Position,
  Order,
  OrderParams,
} from "./types";

const BASE_URL = "https://indexer.dydx.trade/v4";

/**
 * dYdX v4 exchange adapter.
 * Uses the dYdX v4 Indexer REST API.
 * Docs: https://docs.dydx.exchange/
 */
export class DydxAdapter implements ExchangeAdapter {
  readonly id = "dydx" as const;
  readonly name = "dYdX";
  readonly chainId = 1;

  private address: string | null = null;
  private connected = false;

  async connect(credentials: ExchangeCredentials): Promise<void> {
    this.address = credentials.walletAddress ?? credentials.apiKey ?? null;
    this.connected = true;
  }

  disconnect(): void {
    this.address = null;
    this.connected = false;
  }

  isConnected(): boolean {
    return this.connected;
  }

  async getMarkets(): Promise<MarketInfo[]> {
    const res = await fetch(`${BASE_URL}/perpetualMarkets`);
    const data = await res.json();
    const markets = data.markets ?? {};
    return (Object.values(markets) as Record<string, unknown>[]).map((m) => ({
      symbol: m.ticker as string,
      baseAsset: (m.ticker as string).replace("-USD", ""),
      quoteAsset: "USD",
      minSize: Number(m.stepBaseQuantums ?? 0),
      tickSize: Number(m.subticksPerTick ?? 0.01),
      maxLeverage: 20,
      fundingRate: 0,
      openInterest: Number(m.openInterest ?? 0),
      volume24h: Number(m.volume24H ?? 0),
      lastPrice: Number(m.oraclePrice ?? 0),
    }));
  }

  async getCandles(symbol: string, interval: string, limit = 100): Promise<Candle[]> {
    const resolution = this.mapInterval(interval);
    const res = await fetch(
      `${BASE_URL}/candles/perpetualMarkets/${symbol}?resolution=${resolution}&limit=${limit}`
    );
    const data = await res.json();
    return (data.candles ?? []).map((c: Record<string, unknown>) => ({
      timestamp: new Date(c.startedAt as string).getTime(),
      open: Number(c.open),
      high: Number(c.high),
      low: Number(c.low),
      close: Number(c.close),
      volume: Number(c.baseTokenVolume ?? 0),
    }));
  }

  async getFundingRates(): Promise<FundingRate[]> {
    const markets = await this.getMarkets();
    const rates: FundingRate[] = [];
    // Batch first 10 markets for funding rates
    for (const m of markets.slice(0, 10)) {
      try {
        const res = await fetch(
          `${BASE_URL}/historicalFunding/${m.symbol}?limit=1`
        );
        const data = await res.json();
        const latest = data.historicalFunding?.[0];
        rates.push({
          symbol: m.symbol,
          rate: latest ? Number(latest.rate) : 0,
          nextFundingTime: Date.now() + 3600000,
          predictedRate: null,
        });
      } catch {
        rates.push({
          symbol: m.symbol,
          rate: 0,
          nextFundingTime: Date.now() + 3600000,
          predictedRate: null,
        });
      }
    }
    return rates;
  }

  async getOrderbook(symbol: string, depth = 20): Promise<OrderbookSnapshot> {
    const res = await fetch(`${BASE_URL}/orderbooks/perpetualMarket/${symbol}`);
    const data = await res.json();
    return {
      symbol,
      bids: (data.bids ?? []).slice(0, depth).map((b: Record<string, string>) => [Number(b.price), Number(b.size)]),
      asks: (data.asks ?? []).slice(0, depth).map((a: Record<string, string>) => [Number(a.price), Number(a.size)]),
      timestamp: Date.now(),
    };
  }

  async getBalance(): Promise<AccountBalance> {
    this.requireAuth();
    const res = await fetch(`${BASE_URL}/addresses/${this.address}/subaccounts`);
    const data = await res.json();
    const sub = data.subaccounts?.[0];
    if (!sub) {
      return { equity: 0, freeCollateral: 0, totalMarginUsed: 0, unrealizedPnl: 0, realizedPnl: 0 };
    }
    return {
      equity: Number(sub.equity ?? 0),
      freeCollateral: Number(sub.freeCollateral ?? 0),
      totalMarginUsed: Number(sub.equity ?? 0) - Number(sub.freeCollateral ?? 0),
      unrealizedPnl: 0,
      realizedPnl: 0,
    };
  }

  async getPositions(): Promise<Position[]> {
    this.requireAuth();
    const res = await fetch(
      `${BASE_URL}/addresses/${this.address}/subaccounts/0/perpetualPositions?status=OPEN`
    );
    const data = await res.json();
    return (data.positions ?? []).map((p: Record<string, unknown>) => ({
      id: `${p.market}-${this.address}`,
      symbol: p.market as string,
      side: p.side === "LONG" ? "long" : "short",
      size: Math.abs(Number(p.size ?? 0)),
      entryPrice: Number(p.entryPrice ?? 0),
      markPrice: 0,
      liquidationPrice: null,
      unrealizedPnl: Number(p.unrealizedPnl ?? 0),
      realizedPnl: Number(p.realizedPnl ?? 0),
      leverage: 1,
      margin: 0,
      timestamp: new Date(p.createdAt as string).getTime(),
    }));
  }

  async getOpenOrders(): Promise<Order[]> {
    this.requireAuth();
    const res = await fetch(
      `${BASE_URL}/addresses/${this.address}/subaccounts/0/orders?status=OPEN`
    );
    const data = await res.json();
    return (data ?? []).map((o: Record<string, unknown>) => ({
      id: o.id as string,
      symbol: o.ticker as string,
      side: o.side === "BUY" ? "buy" : "sell",
      type: o.type === "LIMIT" ? "limit" : "market",
      size: Number(o.size ?? 0),
      price: Number(o.price ?? 0),
      filledSize: Number(o.totalFilled ?? 0),
      status: "open" as const,
      timestamp: new Date(o.createdAtHeight as string).getTime(),
    }));
  }

  async getOrderHistory(limit = 50): Promise<Order[]> {
    this.requireAuth();
    const res = await fetch(
      `${BASE_URL}/fills?address=${this.address}&subaccountNumber=0&limit=${limit}`
    );
    const data = await res.json();
    return (data.fills ?? []).map((f: Record<string, unknown>) => ({
      id: f.id as string,
      symbol: f.market as string,
      side: f.side === "BUY" ? "buy" : "sell",
      type: f.type === "LIMIT" ? "limit" : "market",
      size: Number(f.size ?? 0),
      price: Number(f.price ?? 0),
      filledSize: Number(f.size ?? 0),
      status: "filled" as const,
      timestamp: new Date(f.createdAt as string).getTime(),
    }));
  }

  async placeOrder(_params: OrderParams): Promise<Order> {
    this.requireAuth();
    throw new Error(
      "dYdX v4 order placement requires cosmos signing. Use the trading UI with connected wallet."
    );
  }

  async cancelOrder(_orderId: string): Promise<void> {
    this.requireAuth();
    throw new Error("Cancel requires cosmos signing. Use the trading UI.");
  }

  async cancelAllOrders(): Promise<void> {
    this.requireAuth();
    throw new Error("Cancel requires cosmos signing. Use the trading UI.");
  }

  private requireAuth() {
    if (!this.connected || !this.address) {
      throw new Error("Connect wallet first");
    }
  }

  private mapInterval(interval: string): string {
    const map: Record<string, string> = {
      "1m": "1MIN",
      "5m": "5MINS",
      "15m": "15MINS",
      "30m": "30MINS",
      "1h": "1HOUR",
      "4h": "4HOURS",
      "1d": "1DAY",
    };
    return map[interval] ?? "1HOUR";
  }
}
